File: sdkwork-rtc-sdk-java/src/main/java/com/sdkwork/rtc/standard/RtcStandardContract.java
package com.sdkwork.rtc.standard;

public final class RtcStandardContract {

  private RtcStandardContract() {
  }

  public interface RtcProviderDriver<TNativeClient> {
    String providerKey();

    RtcClient<TNativeClient> createClient();
  }

  public interface RtcDriverManager<TNativeClient> {
    RtcProviderDriver<TNativeClient> resolveDriver(String providerKey);
  }

  public interface RtcDataSource<TNativeClient> {
    RtcClient<TNativeClient> createClient();
  }

  public interface RtcClient<TNativeClient> {
    void join();

    void leave();

    void publish(String trackId);

    void unpublish(String trackId);

    void muteAudio(boolean muted);

    void muteVideo(boolean muted);

    TNativeClient unwrap();
  }

  public interface RtcRuntimeController<TNativeClient> {
    void join();

    void leave();

    void publish(String trackId);

    void unpublish(String trackId);

    void muteAudio(boolean muted);

    void muteVideo(boolean muted);
  }
}

File: sdkwork-rtc-sdk-kotlin/src/main/kotlin/com/sdkwork/rtc/standard/RtcStandardContract.kt
package com.sdkwork.rtc.standard

object RtcStandardContract {
    const val SYMBOL: String = "RtcStandardContract"
}

interface RtcProviderDriver<TNativeClient> {
    val providerKey: String

    fun createClient(): RtcClient<TNativeClient>
}

interface RtcDriverManager<TNativeClient> {
    fun resolveDriver(providerKey: String): RtcProviderDriver<TNativeClient>
}

interface RtcDataSource<TNativeClient> {
    fun createClient(): RtcClient<TNativeClient>
}

interface RtcClient<TNativeClient> {
    fun join()

    fun leave()

    fun publish(trackId: String)

    fun unpublish(trackId: String)

    fun muteAudio(muted: Boolean)

    fun muteVideo(muted: Boolean)

    fun unwrap(): TNativeClient?
}

interface RtcRuntimeController<TNativeClient> {
    fun join()

    fun leave()

    fun publish(trackId: String)

    fun unpublish(trackId: String)

    fun muteAudio(muted: Boolean)

    fun muteVideo(muted: Boolean)
}
